File: ml-benchmark/synduce/list/bal.f
Config ExtraGrammar = "AutoLifter";
Inductive List = nil Unit | cons {Int, List};
Inductive CList = cnil Unit | single Int | concat {CList, CList};

min = \a: Int. \b: Int. if > a b then b else a;

spec = \xs: List. (fix (
  \f: List -> {Int, Int, Bool}. \x: List.
  match x with
    nil _ -> {0, 0, true}
  | cons {h, t} ->
    let result = f t in
    let cnt = result.1 in
    let min_cnt = result.2 in
    let bal = result.3 in
    let new_cnt = if (> h 0) then (+ cnt 1) else (- cnt 1) in
    {new_cnt, min min_cnt new_cnt, and bal (>= new_cnt 0)}
  end
) xs).3;

cat = fix (
  \f: List -> List -> List. \a: List. \b: List.
  match a with
    nil _ -> b
  | cons {h, t} -> cons {h, (f t b)}
  end
);

repr = fix (
    \f: CList -> Compress List. \xs: CList.
    match xs with
      cnil _ -> nil unit
    | single x -> cons {x, nil unit}
    | concat {a, b} -> cat (f a) (f b)
    end
);

main = \x: CList. spec (repr x);

File: ml-benchmark/synduce/list/bal.ml
type list =
  | Nil
  | Cons of int * list

type clist =
  | CNil
  | Single of int
  | Concat of clist * clist

let min a b = if a > b then b else a

let spec xs =
  let rec f x =
    match x with
    | Nil -> (0, 0, true)
    | Cons (h, t) ->
      match f t with
      | (cnt, min_cnt, bal) ->
        let new_cnt = if h > 0 then cnt + 1 else cnt - 1 in
        (new_cnt, min min_cnt new_cnt, bal && new_cnt >= 0)
  in
  match f xs with
  | (_, _, result) -> result

let cat =
  let rec f a b =
    match a with
    | Nil -> b
    | Cons (h, t) -> Cons (h, f t b)
  in
  f

val repr: clist -> list compress
let repr =
  let rec f xs =
    match xs with
    | CNil -> Nil
    | Single x -> Cons (x, Nil)
    | Concat (a, b) -> cat (f a) (f b)
  in
  f

let prog x = spec (repr x)